src/backend/pipeline/cont_combiner.c

```diff
@@ -488,21 +494,26 @@ sync_combine(ContQueryCombinerState *state, Tuplestorestate *results, TupleHashT
 			IncrementCQWrite(1, HEAPTUPLESIZE + slot->tts_tuple->t_len);
 		}
 
+		if (state->cache)
+			cache_tuples = lappend(cache_tuples, ExecCopySlotTuple(slot));
+
 		ResetPerTupleExprContext(estate);
 	}
 
 	/*
 	 * We need to wait until we've completed all updates/inserts before caching everything, otherwise
-	 * we may free a cached tuple before trying to read it
+	 * we may free a cached tuple before trying to read it.
 	 */
-	if (state->cache)
+	foreach(lc, cache_tuples)
 	{
-		foreach_tuple(slot, results)
-		{
-			GroupCachePut(state->cache, slot);
-		}
+		HeapTuple tup = (HeapTuple) lfirst(lc);
+		ExecStoreTuple(tup, slot, InvalidBuffer, false);
+		GroupCachePut(state->cache, slot);
+		heap_freetuple(tup);
```

Comment:
I just try to free up any in-memory tuples eagerly since they can hog a lot of memory even if for a short duration.